### PR TITLE
Add new ELPA support

### PR DESCRIPTION
**Pull request by pplab**
ABACUS only supported ELPA which is newer than version 2017 the default ELPA kernel is not the one that is optimized for modern CPU families.

We add a GenELPA package, which is an easy-to-use ELPA interface to solve generalized eigenvalue problems. With this package, ABACUS now supports all versions of ELPA instead of the new ones. Also, the program will check the hardware and choose the better (if not the best) kernel automatically, so it will run (much) faster.

We use automatic scripts to handle the ELPA library of different versions, therefore the installation and usage remain the same.

### CODE REVIEW

File: source/module_hsolver/genelpa/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+#PROJECT (TEST)
+
+#SET(ELPA_INCLUDE_DIR /home/nic/yshen/program/elpa/2016.05.004/include/elpa-2016.05.004/)
+#SET(ELPA_INCLUDE_DIR /home/nic/yshen/program/elpa/2021.11.002-A100/include/elpa-2021.11.002)
+MESSAGE(STATUS "elpa include dir:" ${ELPA_INCLUDE_DIR} )
+string(REGEX REPLACE "/$" "" ELPA_INC_STR ${ELPA_INCLUDE_DIR})
+string(REPLACE "/" ";" ELPA_INC_STR ${ELPA_INC_STR})
+list(GET ELPA_INC_STR -1 ELPA_VER)
+string(REPLACE "elpa-" "" ELPA_VER ${ELPA_VER})
+message(STATUS "elpa version is: " ${ELPA_VER})
```
Comment (caic99) [inline source/module_hsolver/genelpa/CMakeLists.txt:11]: Is it possible to get the elpa version from header file, like [modules/FindELPA.cmake](https://github.com/abacusmodeling/abacus-develop/pull/171/files#diff-b99d60b24a7770488bf35977dfad7f653142eeabd7964896f3ba5361ca97a76a)?

File: source/module_hsolver/genelpa/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+if(${ELPA_VER} VERSION_GREATER "2016.05.004")
+    execute_process(COMMAND /bin/sh ${CMAKE_CURRENT_SOURCE_DIR}/translate_elpa_generic.sh ${ELPA_INCLUDE_DIR})
+    add_library(genelpa OBJECT elpa_new.cpp elpa_new_real.cpp elpa_new_complex.cpp utils.cpp)
+else()
+    execute_process(COMMAND /bin/sh ${CMAKE_CURRENT_SOURCE_DIR}/translate_elpa_generated.sh ${ELPA_INCLUDE_DIR})
```
Comment (caic99) [inline source/module_hsolver/genelpa/CMakeLists.txt:17]: Is it possible to use macros to determine and select the correct interface for elpa, rather than use plain-text replacing?

Comment (pplab) [reply source/module_hsolver/genelpa/CMakeLists.txt:17]: the header files of ELPA follow C standards but some features are not well compatible with C++.

For example, the `_Generic` keyword should be reimpliment in C++ by overload functions.

Another one is the `complex` keyword, defined in C's standard header `complex.h`, is not compatible with C++. One can use `#define complex _Complex` to resolve this problem, but this does NOT work every time for every versions of compiler.

In short, The ELPA does not provide the "correct" interfaces working with standard C++, therefore it's better to create the new header files for C++ rather than try to make the C header files work with C++.

File: source/module_hsolver/genelpa/CMakeLists.txt (hunk 1 above)
Comment (caic99) [reply source/module_hsolver/genelpa/CMakeLists.txt:11]: > in https://github.com/abacusmodeling/abacus-develop/pull/171/files/4774d4492d2ec192f807ca034ced9d0a2ae6b8dd#diff-6d9fd3d243dafc589abb0ed35cbc82bb2b9f36ade9fc2ec91956c45ab8fff267

This method extracts elpa version info from include file path. However, those information won't be always available. The elpa installed from package management system does not contain version in folder name, e.g. `/usr/include/elpa/elpa/elpa.h`.
Personally, I think this is a dirty workaround. The version of elpa should be retrieved from `elpa_version.h`, but I'm not 100% sure if the legacy elpa contains this file.

File: source/module_hsolver/genelpa/CMakeLists.txt (hunk 2 above)
Comment (caic99) [reply source/module_hsolver/genelpa/CMakeLists.txt:17]: @pplab Sorry, let me explain my concern again:
What I'm suggesting is avoid duplicated `elpa_generic_template_*.hpp`. Personally I think in the form of 
```c++
#include <elpa_version.h>
static inline void elpa_eigenvalues(elpa_t handle, float complex *a, float *ev, int *error)
{
#if ELPA_API_VERSION >= 20211125
	elpa_eigenvalues_fc(handle, a, ev, error);
#else 
	elpa_eigenvalues_all_host_arrays_d(handle, a, ev, error);
}
```
should be better. What do you think?

Comment (caic99) [reply source/module_hsolver/genelpa/CMakeLists.txt:17]: The complex compatibility between c and c++ is really bad. I tried to figure out a more general solution but failed, unless compiled separately. Let's leave it as is.

Comment (pplab) [reply source/module_hsolver/genelpa/CMakeLists.txt:17]: > 
This approach would certainly work, and the code would be more streamlined.

There are two reasons for not using this approach. First, I don't know if ELPA team will make major (e.g. structural) changes to this header file in the future to make this conditional compilation less easy to modify; second, inserting a lot of conditional compilation instructions into the code would make me feel like I am always interrupted when reading the code, so I personally don't like this approach.

**Comment by caic99** [comment]
@pplab Thank you for your contribution! However, this PR lacks the support to Makefile, and it does not contain necessary documentation change(i.e. the required version of elpa, etc.).
